substrate/frame/system/src/extensions/check_nonce.rs

```diff
@@ -58,83 +61,87 @@ impl<T: Config> core::fmt::Debug for CheckNonce<T> {
-	type Call = T::RuntimeCall;
-	type AdditionalSigned = ();
+	type Val = Option<(T::AccountId, AccountInfo<T::Nonce, T::AccountData>)>;
 	type Pre = ();
-	const IDENTIFIER: &'static str = "CheckNonce";
-
-	fn additional_signed(&self) -> core::result::Result<(), TransactionValidityError> {
-		Ok(())
-	}
-
-	fn pre_dispatch(
-		self,
-		who: &Self::AccountId,
-		_call: &Self::Call,
-		_info: &DispatchInfoOf<Self::Call>,
-		_len: usize,
-	) -> Result<(), TransactionValidityError> {
-		let mut account = crate::Account::<T>::get(who);
-		if account.providers.is_zero() && account.sufficients.is_zero() {
-			// Nonce storage not paid for
-			return Err(InvalidTransaction::Payment.into())
-		}
-		if self.0 != account.nonce {
-			return Err(if self.0 < account.nonce {
-				InvalidTransaction::Stale
-			} else {
-				InvalidTransaction::Future
-			}
-			.into())
-		}
-		account.nonce += T::Nonce::one();
-		crate::Account::<T>::insert(who, account);
-		Ok(())
-	}
 
 	fn validate(
 		&self,
-		who: &Self::AccountId,
-		_call: &Self::Call,
-		_info: &DispatchInfoOf<Self::Call>,
+		origin: <T as Config>::RuntimeOrigin,
+		_call: &T::RuntimeCall,
+		_info: &DispatchInfoOf<T::RuntimeCall>,
 		_len: usize,
-	) -> TransactionValidity {
+		_context: &mut Context,
+		_self_implicit: Self::Implicit,
+		_inherited_implication: &impl Encode,
+	) -> ValidateResult<Self::Val, T::RuntimeCall> {
+		let Some(who) = origin.as_system_origin_signer() else {
+			return Ok((Default::default(), None, origin))
+		};
 		let account = crate::Account::<T>::get(who);
 		if account.providers.is_zero() && account.sufficients.is_zero() {
 			// Nonce storage not paid for
-			return InvalidTransaction::Payment.into()
+			return Err(InvalidTransaction::Payment.into())
 		}
 		if self.0 < account.nonce {
-			return InvalidTransaction::Stale.into()
+			return Err(InvalidTransaction::Stale.into())
 		}
 
-		let provides = vec![Encode::encode(&(who, self.0))];
+		let provides = vec![Encode::encode(&(who.clone(), self.0))];
 		let requires = if account.nonce < self.0 {
-			vec![Encode::encode(&(who, self.0 - One::one()))]
+			vec![Encode::encode(&(who.clone(), self.0.saturating_sub(One::one())))]
 		} else {
 			vec![]
 		};
 
-		Ok(ValidTransaction {
+		let validity = ValidTransaction {
 			priority: 0,
 			requires,
 			provides,
 			longevity: TransactionLongevity::max_value(),
 			propagate: true,
-		})
+		};
+
+		Ok((validity, Some((who.clone(), account)), origin))
+	}
+
+	fn prepare(
+		self,
+		val: Self::Val,
+		_origin: &T::RuntimeOrigin,
+		_call: &T::RuntimeCall,
+		_info: &DispatchInfoOf<T::RuntimeCall>,
+		_len: usize,
+		_context: &Context,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let Some((who, mut account)) = val else { return Ok(()) };
+		// `self.0 < account.nonce` already checked in `validate`.
+		if self.0 > account.nonce {
+			return Err(InvalidTransaction::Future.into())
+		}
+		account.nonce += T::Nonce::one();
+		crate::Account::<T>::insert(who, account);
```

Comment:
Here we assume that the account information queried in validate has not been touched by the preparation of any prior transaction extension in the pipeline.

It feels a bit error prone.

Moreover `AccountInfo` contains a generic `AccountData` that other pallet may utilize.

Maybe have a warning in bold in the doc or just fetch the storage again.

Reply:
> Here we assume that the account information queried in validate has not been touched by the preparation of any prior transaction extension in the pipeline.
>
> It feels a bit error prone.

It's true that we make this assumption, but I think it's a fair assumption because what else (other than a `frame_system` mechanism i.e. `CheckNonce`) should ever mutate the nonce? If someone ever does do anything like that, they would either not use `CheckNonce` altogether or else their transactions would never work. I will document this assumption in the code anyway, just to be safe.

> Moreover `AccountInfo` contains a generic `AccountData` that other pallet may utilize.

I think this is a good callout because it's inexpensive for us to pass just the nonce in `Val` and do a `mutate` instead of an `insert` in the `Account` map.

Reply:
Perfect